src/api-service/__app__/onefuzzlib/job_templates/render.py

```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/env python
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+import json
+from typing import Dict, List
+
+from jsonpatch import apply_patch
+from memoization import cached
+from onefuzztypes.enums import ContainerType, UserFieldType
+from onefuzztypes.job_templates import (
+    TEMPLATE_BASE_FIELDS,
+    JobTemplate,
+    JobTemplateConfig,
+    JobTemplateField,
+    JobTemplateRequest,
+    TemplateUserData,
+    UserField,
+)
+
+
+def template_container_types(template: JobTemplate) -> List[ContainerType]:
+    return list(set(y.type for x in template.tasks for y in x.containers if not y.name))
```

Comment:
Nit, binding to `t` instead of `x` and `c` instead of `y` would be a helpful mnemonic scheme.